src/zenml/zen_stores/sql_zen_store.py

```diff
@@ -11449,119 +11557,327 @@ def update_tag(
+                ),
+            )
+            .where(TagResourceSchema.tag_id == tag.id)
+            .group_by(scope_id_column)
+        )
+        results = session.exec(query).all()
+        conflicts = [k for k, v in results if v > 1]
+        if conflicts:
+            error_message = (
+                f"The tag `{tag.name}` is associated with multiple entries of "
+                f"`{resource_type.value}`s"
+            )
 
-    def _get_tag_model_schema(
+            if resource_id_column != scope_id_column:
+                error_message += (
+                    f" that share the same `{scope_id_column.name}`s"
+                )
+            error_message += f": {conflicts}"
+            return False, error_message
+
+        return True, ""
+
+    def _get_tag_resource_schema(
         self,
-        tag_id: UUID,
-        resource_id: UUID,
-        resource_type: TaggableResourceTypes,
+        tag_resource: TagResourceRequest,
         session: Session,
     ) -> TagResourceSchema:
-        """Gets a tag model schema by tag and resource.
+        """Gets a tag model schema.
 
         Args:
-            tag_id: The ID of the tag to get.
-            resource_id: The ID of the resource to get.
-            resource_type: The type of the resource to get.
+            tag_resource: The tag resource to get.
             session: The database session to use.
 
         Returns:
             The tag resource schema.
 
         Raises:
-            KeyError: if entity not found.
+            KeyError: if a relationship between the tag and the resource does
+                not exist.
         """
         schema = session.exec(
             select(TagResourceSchema).where(
-                TagResourceSchema.tag_id == tag_id,
-                TagResourceSchema.resource_id == resource_id,
-                TagResourceSchema.resource_type == resource_type.value,
+                TagResourceSchema.tag_id == tag_resource.tag_id,
+                TagResourceSchema.resource_id == tag_resource.resource_id,
+                TagResourceSchema.resource_type
+                == tag_resource.resource_type.value,
             )
         ).first()
         if schema is None:
             raise KeyError(
-                f"Unable to get {TagResourceSchema.__tablename__} with IDs "
-                f"`tag_id`='{tag_id}' and `resource_id`='{resource_id}' and "
-                f"`resource_type`='{resource_type.value}': No "
-                f"{TagResourceSchema.__tablename__} with these IDs found."
+                f"Tag `{tag_resource.tag_id}` is not currently assigned to "
+                f"{tag_resource.resource_type.value} with ID "
+                f"`{tag_resource.resource_id}`."
             )
         return schema
 
-    def _create_tag_resource(
-        self, tag_resource: TagResourceRequest, session: Session
-    ) -> TagResourceResponse:
-        """Creates a new tag resource relationship.
+    def _create_tag_resource_schemas(
```

Comment:
The main issue I have with this (and it's a big one) is race conditions: It follows the usual
try to fetch first, if it doesn't exist create it paradigm which is certain to lead to race conditions on hugely parallel requests.

Even worse, the whole exclusivity checks are also prone to race conditions, making the feature not really realiable.

Reply:
@stefannica and I implemented a new system for the tags and tag resources to avoid race conditions, let us know what you think.